### DISCUSSION TITLE
请教一下关于C++注册非DYNAMIC_DELEGATE回调的问题

### DISCUSSION THREAD

**Discussion by watsonsong** [Tencent/puerts#1486] [Q&A]
我看[文档](https://puerts.github.io/docs/puerts/unreal/engine_call_script)中到注册非DYNAMIC_DELEGATE有几种方法，包括：
* 使用FCallback参数配合toDelegate；
* 使用std::function，看示例似乎不需要toDelegate；
* 使用[FJsObject](https://puerts.github.io/docs/puerts/unreal/template_binding#js%E5%87%BD%E6%95%B0%E6%98%A0%E5%B0%84jsobject%E5%B9%B6%E5%9B%9E%E8%B0%83)来做回调

使用std::function注册的时候，发现如果构造闭包传递给C++，然后通过AddLambda的方式注册给一个非DYNAMIC的回调。延迟一段时间后再触发回调的话，比较容易产生崩溃；感觉是C++持有的std::function对象并不会持有这个闭包的生命周期。

Unreal中有很多非DYNAMIC的事件(特别是GAS中)，它们通过各种AbilityAsync让蓝图能调用。因此想非DYNAMIC的回调能否安全的通过C++绑定封装给JS。
否则的话得为这些每个回调再封装一个DYNAMIC_DELEGATE，类似[ULatentActionState](https://github.com/chexiongsheng/puerts_unreal_demo/blob/41d1eb5af1abb84270bf6511c354fa88d5c2287c/Source/puerts_unreal_demo/LatentActionState.h#L15)这样再来做一个代理中转，总感觉会比较麻烦而且有损性能。

不知道能否JS传递一个闭包给到C++，然后给注册给一个非Dynamic的Delegate，能持有或者不持有(AddWeak)生命周期的方法？

**Comment by chexiongsheng** [comment]
目前提供的所有c++调用脚本的方式传闭包都会持有闭包的引用不会释放，这是最最基本的东西。
std::function还加了对JsEnv的生命周期跟踪，直接使用std::function即使虚拟机释放了调用也没问题。包括puerts自己内部也在用std::function。

**Comment by watsonsong** [comment]
测试了一下发现我描述有问题，产生问题的原因还和FJsObject有关系。在使用UMG的ListView的时候为了方便，包装了一个统一的UObject去持有一个FJsObject。这样在AddItem的时候能直接Add一个纯粹的JS对象进去。
然后在一些监听事件（例如选中状态的改变）里面，希望把形式封装成直接拿到一个JS对象回来。因此C++侧做了一个封装，回调的UObject转回一个FJsObject再传递给JS。


大概代码如下。
```
UCLASS(BlueprintType)
class JSENV_API UJsObjectWrapper : public UObject
{
	GENERATED_UCLASS_BODY()

public:
	UPROPERTY(Transient, BlueprintReadWrite)
	FJsObject JsObject;

        // 其他在puerts关闭的时候，清理所有UJsObjectWrapper的操作
};

static FDelegateHandle ListenItemJSSelectionChanged(UListView& ListView,
	const std::function<void(const FJsObject&, bool)>& OnItemSelection)
{
	return ListView.OnItemSelectionChanged().AddLambda([OnItemSelection](UObject* Item)
	{
		if (Item == nullptr)
		{
			OnItemSelection(FJsObject(), false);
			return;
		}

		if (const UJsObjectWrapper* Holder = Cast<UJsObjectWrapper>(Item))
		{
			OnItemSelection(Holder->JsObject, true);
		}
		else
		{
			UE_LOG(LogTemp, Warning, TEXT("The list view selected item is not a UJsObjectWrapper."));
		}
	});
}
```

然后崩溃发生在这个std::function被Invoke的时候，FJsObject拷贝的过程中。`FJsObject(const FJsObject& InOther)`这个拷贝构造函数，在InOther.GContext.Get(Isolate)的时候发生的崩溃。
不太清楚是否应该这么用，因为如果这个封装发生在js侧，由JS拿到这个UObject，再转换成UJsObjectWrapper这个类型，获取JsObject则不会有任何问题。看起来UJsObjectWrapper本身的生命周期应该是在的。

想求教一下是不是使用和理解上有什么问题，导致了这个崩溃的产生。

**Comment by chexiongsheng** [comment]
你试试更新下：https://github.com/Tencent/puerts/commit/67615a1f73eb902347b71896802dd2570d1a6ffa
